### PR TITLE
Add support for connecting over abstract unix sockets

### PR DESCRIPTION
**Pull request by psychon**
This adds support for connecting over abstract unix sockets. The
connection is established via functions provided by nix. The resulting
socket is wrapped in a RawFdContainer so that it is automatically closed
on drop.

All the actual I/O already uses as_raw_fd() to access the RawFd and work
directly with that. Thus, the existing code for this just works.

In the future, when there is a safe way to construct a UnixStream from a
RawFd, the AbstractUnixStream case that this commit adds should be be
removed again (e.g. nix could add wrappers using OwnedFd, when that type
becomes stable).

Signed-off-by: Uli Schlachter <psychon@znc.in>

### CODE REVIEW

**Comment by psychon** [comment]
> I like this. My only issue is that it should have more tests to handle the new abstract socket use cases.

All the testing that I did so far was "run under `strace`". However, I didn't see any difference between abstract / concrete unix socket, so I also added some `println` to check that this code is really executed. That's it.

>  If you don't have time, I could go and add tests if you wanted to.

I'm not sure, actually. Feel free to do what you have in mind for this.

File: x11rb/src/rust_connection/stream.rs [hunk 1]
```diff
@@ -183,6 +185,8 @@ enum DefaultStreamInner {
     TcpStream(TcpStream),
     #[cfg(unix)]
     UnixStream(UnixStream),
+    #[cfg(unix)]
+    AbstractUnix(RawFdContainer),
```
[the remark this answers is not in the document]
Comment (psychon) [reply x11rb/src/rust_connection/stream.rs:189]: Actually, we can remove the `UnixStream`-case. I think we can convert a `UnixStream` to a `RawFdContainer` in safe code via the `IntoRawFd` trait. Only downside would be: The `peer_addr()` implementation would no longer be possible the way it is right now. Hm. :-(

Dunno if it is really worth adding `socket2` for this. Compile times and such....

File: x11rb/src/rust_connection/stream.rs [hunk 2]
```diff
@@ -196,12 +200,19 @@ impl DefaultStream {
...
             ConnectAddress::Socket(path) => {
-                // TODO: Try abstract socket (file name with prepended '\0')
-                // Not supported on Rust right now: https://github.com/rust-lang/rust/issues/42048
-
-                // connect over Unix domain socket
-                let stream = UnixStream::connect(path)?;
-                Self::from_unix_stream(stream)
+                // Try abstract unix socket first. If that fails, fall back to normal unix socket
+                if let Ok(stream) = connect_abstract_unix_stream(path.as_os_str().as_bytes()) {
+                    // TODO: Does it make sense to add a constructor similar to from_unix_stream()?
```
[the remark this answers is not in the document]
Comment (psychon) [reply x11rb/src/rust_connection/stream.rs:205]: So I guess we should at least wait until the I/O safety work in Rust is stabilised. That should provide some possibilities (and also some guidelines on how to use these possibilities).

File: x11rb/src/rust_connection/stream.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (notgull) [reply x11rb/src/rust_connection/stream.rs:189]: Yeah, it's definitely a puzzle. I'd say to leave it the way it is for now.

**Review by notgull — APPROVED** [review]

**Pull request merged**